05-AdvancedTopics/web-search-mcp/README.md

```diff
@@ -0,0 +1,214 @@
+# Web Search MCP Server
+
+This advanced sample demonstrates how to build a sophisticated MCP server and client using FastMCP to extend LLM capabilities with real-time web data via SerpAPI, a critical skill for developing dynamic AI agents.
+
+## Overview
+
+This implementation features four tools that showcase MCP's ability to handle diverse, external API-driven tasks securely and efficiently:
+
+- **general_search**: For broad web results
+- **news_search**: For recent headlines
+- **product_search**: For e-commerce data
+- **qna**: For question-and-answer snippets
+
+## Features
+
+- **External API Integration**: Demonstrates secure handling of API keys and external requests
+- **Structured Data Parsing**: Shows how to transform API responses into LLM-friendly formats
+- **Error Handling**: Robust error handling with appropriate logging
+- **Interactive Client**: Includes both automated tests and an interactive mode for testing
+- **Context Management**: Leverages MCP Context for logging and tracking requests
+
+## Prerequisites
+
+- Python 3.8 or higher
+- SerpAPI API Key (Sign up at [SerpAPI](https://serpapi.com/) - free tier available)
+
+## Installation
+
+1. Install dependencies using uv (recommended) or pip:
+
+```bash
+# Using uv (recommended)
+uv pip install -r requirements.txt
+
+# Using pip
+pip install -r requirements.txt
+```
+
+2. Create a `.env` file in the project root with your SerpAPI key:
+
+```
+SERPAPI_KEY=your_serpapi_key_here
+```
+
+## Usage
+
+### Running the Server
+
+Start the MCP server:
+
+```bash
+python server.py
+```
+
+The server will run as a stdio-based MCP server that the client can connect to directly.
+
+### Running the Client
+
+Run the automated tests (this will automatically start the server):
+
+```bash
+python client.py
+```
+
+Or run in interactive mode:
+
+```bash
+python client.py --interactive
+```
+
+### Testing with Different Methods
+
+#### Using the Interactive Mode
+The easiest way to test the tools is using the interactive mode:
+
+```bash
+python client.py --interactive
+```
+
+This provides a menu-driven interface to try each tool with custom queries.
+
+#### Using the MCP Python SDK
+You can also build your own test scripts using the MCP Python SDK:
+
+```python
+from mcp import ClientSession, StdioServerParameters
+from mcp.client.stdio import stdio_client
+
+async def test_custom_query():
+    server_params = StdioServerParameters(
+        command="python",
+        args=["server.py"],
+    )
+    
+    async with stdio_client(server_params) as (reader, writer):
+        async with ClientSession(reader, writer) as session:
+            await session.initialize()
+            # Call tools with your custom parameters
+            result = await session.call_tool("general_search", 
+                                           arguments={"query": "your custom query"})
+            # Process the result
+```
+
+## Tool Descriptions
```

Comment:
Add lead in sentence here, there shouldn't be an h3 header directly after an h2.